GoogleContainerTools/skaffold
# Show update check to users once in 24 hours

Update check prompts are shown on every command run. 
This could annoy the users and get them to switch off update checks.

In this PR
1) Add a config section `update-check` with `last-prompted` field.
2) Add logic to display update check prompt only once in 24 hours.

This does not change the existing update check behavior. See #5741 for that.

Testing done:

1) Build a binary with latest version & make sure 
 a) update check message is not shown. 
 b) `update-config` is not updated with `last-prompted`

```
GOOS=darwin GOARCH=amd64 CGO_ENABLED=1 \
            go build -gcflags="all=-N -l" -tags "release " -ldflags "-X github.com/GoogleContainerTools/skaffold/pkg/skaffold/version.version=v1.24.0 -s -w " -o out/skaffold github.com/GoogleContainerTools/skaffold/cmd/skaffold
```

```
.➜  multi-config-microservices git:(show_update_check_dailu) ✗ ../../out/skaffold build -d gcr.io/tejal-gke1
Generating tags...
 - base -> gcr.io/tejal-gke1/base:v1.23.0-36-gbf79f0788
 - leeroy-app -> gcr.io/tejal-gke1/leeroy-app:v1.23.0-36-gbf79f0788
 - leeroy-web -> gcr.io/tejal-gke1/leeroy-web:v1.23.0-36-gbf79f0788
Checking cache...
 - base: Found Remotely
 - leeroy-app: Found Remotely
 - leeroy-web: Found Remotely                  

➜  multi-config-microservices git:(show_update_check_dailu) ✗ cat ~/.skaffold/config
global:
  survey:
    last-prompted: "2021-04-27T13:32:59-07:00"
  collect-metrics: true
kubeContexts: []
```


2) Build a binary with older version  & make sure
a)  update message is shown
```
➜ GOOS=darwin GOARCH=amd64 CGO_ENABLED=1 \
            go build -gcflags="all=-N -l" -tags "release " -ldflags "-X github.com/GoogleContainerTools/skaffold/pkg/skaffold/version.version=v1.22.0 -s -w " -o out/skaffold github.com/GoogleContainerTools/skaffold/cmd/skaffold

➜ ../../out/skaffold build -d gcr.io/tejal-gke1
Generating tags...
 - base -> gcr.io/tejal-gke1/base:v1.23.0-36-gbf79f0788
 - leeroy-app -> gcr.io/tejal-gke1/leeroy-app:v1.23.0-36-gbf79f0788
 - leeroy-web -> gcr.io/tejal-gke1/leeroy-web:v1.23.0-36-gbf79f0788
Checking cache...
 - base: Found Remotely
 - leeroy-app: Found Remotely
 - leeroy-web: Found Remotely
There is a new version (1.23.0) of Skaffold available. Download it from:
  https://github.com/GoogleContainerTools/skaffold/releases/tag/v1.23.0
```
b) `update-config` is updated with `last-prompted`
```
➜ cat ~/.skaffold/config                       
global:
  survey:
    last-prompted: "2021-04-27T13:32:59-07:00"
  collect-metrics: true
  update-config:
    last-prompted: "2021-05-06T00:14:50-07:00"
kubeContexts: []

```
c) subsequent  invocations don't display the update check messages.
```
➜  ../../out/skaffold version && ../../out/skaffold build -d gcr.io/tejal-gke1 
v1.22.0

Generating tags...
 - base -> gcr.io/tejal-gke1/base:v1.23.0-36-gbf79f0788
 - leeroy-app -> gcr.io/tejal-gke1/leeroy-app:v1.23.0-36-gbf79f0788
 - leeroy-web -> gcr.io/tejal-gke1/leeroy-web:v1.23.0-36-gbf79f0788
Checking cache...
 - base: Found Remotely
 - leeroy-app: Found Remotely
 - leeroy-web: Found Remotely
```

d) Change the date to yesterday and make sure update message is  displayed
```
global:
  survey:
    last-prompted: "2021-04-27T13:32:59-07:00"
  collect-metrics: true
  update-config:
    last-prompted: "2021-05-05T00:14:50-07:00"
kubeContexts: []

➜   ../../out/skaffold build -d gcr.io/tejal-gke1 
Generating tags...
 - base -> gcr.io/tejal-gke1/base:v1.23.0-36-gbf79f0788
 - leeroy-app -> gcr.io/tejal-gke1/leeroy-app:v1.23.0-36-gbf79f0788
 - leeroy-web -> gcr.io/tejal-gke1/leeroy-web:v1.23.0-36-gbf79f0788
Checking cache...
 - base: Found Remotely
 - leeroy-app: Found Remotely
 - leeroy-web: Found Remotely
There is a new version (1.23.0) of Skaffold available. Download it from:
  https://github.com/GoogleContainerTools/skaffold/releases/tag/v1.23.0

```
